### ISSUE TITLE
Add support for documentation for Latex commands

### ISSUE THREAD

**Issue by petRUShka** [lervag/vimtex#655]
[LaTeX-Suite](http://vim-latex.sourceforge.net) has a useful feature

> LaTeX help file Latex-suite ships with the standard texinfo file translated into vim help format so that you can get help for common latex commands from within vim. You can view the HTML version of this file here. 

It would be great to have such in vimtex plugin.

Another option is to support [K](http://vimdoc.sourceforge.net/htmldoc/various.html#K) command. That can be also useful.

**Comment by petRUShka** [comment]
Addition information about `K`: it is possible to use `texdoc` utility for example

**Comment by lervag** [comment]
Interesting request. First, regarding `K`: I think `texdoc` only works if you have installed the documentation. I am not 100% sure, but I don't want to use `texdoc` unless it is very reliable. That is, I don't want to respond to a lot of issues about "texdoc doesn't work" or "K mapping doesn't work".

However, if we were to add the [`latexhelp.txt`](https://github.com/vim-latex/vim-latex/blob/master/doc/latexhelp.txt) file from LaTeX-Suite, then we could use the Vim `K` mapping (i.e., use vim help), and it should work pretty well with many commands. I think it might combine quite nicely. I have sent an email to Mikolaj Machowski asking for permission (and for input on the idea of including it).

**Comment by lervag** [comment]
It seems the latexhelp file is based on [this](http://www.ctan.org/tex-archive/info/latex2e-help-texinfo/), which was updated in 2015. It would be more convenient, I think, if we could have a version that was more up to date with regard to this documentation.

Perhaps it is better to instead fork out to a browser, e.g. to the wikibooks? Something like this:
https://en.wikibooks.org/wiki/Special:Search?search=test&prefix=LaTeX%2F&fulltext=Search+this+book&fulltext=Search&searchToken=chh4dh2b2gn13hhgy5b0age1q

**Comment by petRUShka** [comment]
I think both solution is good. But I personally prefer offline one because sometimes It is possible to work with latex somewhere offline.

May be there should be options? For example:

1. Pure offline (`latexrefman`)
2. Pure online
3. Offline with online fallback (in case there isn't help for keyword in `latexrefman`)

Or probably two separate keys for it...

**Comment by lervag** [comment]
How about:

1. `K` to use offline solution through Vims help system with `latexhelp.txt`, and

2. `\ld` (or other) to use online search on wikibooks with default browser?

Both of these would be pretty easy to implement, except number 2 on MacOS and Windows. I guess with some help we would manage that as well.

**Comment by petRUShka** [comment]
`texdoc` isn't an option (at least isn't main option) I agree with you totally. I mentioned it just in case. Moreover probably `texdoc` doesn't support keyword search. Only package based pdf manuals.

May be it is useful for somebody with installed or compiled documentation.

But may be it is also useful to recognize which package is responsible for keyword under cursor and open browser page with corresponding pdf package manual (as fallback in case there isn't texdoc offline documentation).

**Example**: [biber CTAN page](https://www.ctan.org/pkg/biber)  includes [manual in pdf format](http://mirrors.ctan.org/biblio/biber/documentation/biber.pdf).

**Comment by petRUShka** [comment]
Your solution for offline/online keyword help (`K` and `\ld`) looks great!

**Comment by petRUShka** [comment]
But please look also at "open pdf manual" option. Possibly it isn't too much (I see some use cases).

**Comment by poppyschmo** [comment]
Hey guys. Crazy thing: I actually made something for personal use that's relevant to what you've been discussing. Just [uploaded it](https://github.com/poppyschmo/vim-latexrefman). BTW, *latexrefman* is continually updated (see their [svn ](http://svn.gna.org/viewcvs/latexrefman/trunk/) for the latest, if you're interested).

**Comment by petRUShka** [comment]
@poppyschmo thanks for noticing! Last commit to svn was several weeks ago. You are right it should be quite actual.

**Comment by lervag** [comment]
@poppyschmo Could you explain what your plugin does? It is a little bit cryptic to me. Does it take the latest version of the latexrefman and compile a vim help file from it? You mention the license issue. What would happen if I simply included a compiled vim help file?

**Comment by poppyschmo** [comment]
@lervag Apologies for any confusion. I just updated the blurb to now hopefully convey what you correctly gleaned: that my rather unremarkable (excuse for a) plugin simply fetches the latest refman and converts it to a Vim help file.

In regards to the license: I have no clue. I tried emailing the list a while back but never heard anything. Maybe I should have signed with my gpg key (they seem to be all about that sort of thing)? Other than the bit about the deceased author in the license section of the [project page](http://home.gna.org/latexrefman/), I ran across this from the main [about page](http://about.gna.org/#art3):

> For non-program contributions, we will decide on a case per case basis
> whether these restrictions apply or not. For instance, most of the time GPL
> compatibility will not be required for documentation work.

Normally, I'd say screw it, be bold, and ask for forgiveness later. But the Gnu/FSF longbeards scare me, so I'll defer y'all's judgement on this one.

**Comment by kiryph** [comment]
I just want to throw in my 50cent:

There are also following programs offering offline documentation https://zealdocs.org/ and https://kapeli.com/dash (non free, but certainly useful).

The vim plugins https://github.com/keith/investigate.vim or https://github.com/rizzatti/dash.vim map for example `gK` or `<Leader>K` to open the documentation at the right spot: for example if the cursor is on minipage you will get following information presented:

![dash app-minipage](https://cloud.githubusercontent.com/assets/371756/21985496/52f88542-dbfb-11e6-88dd-c5ed1e022523.png)


However, providing vim builtin documentation would of course be a great addition to vimtex.

**Comment by poppyschmo** [comment]
> You mention the license issue. What would happen if I simply included a compiled vim help file?

@lervag The maintainer was nice enough to provide some [clarification](https://mail.gna.org/public/latexrefman-discuss/2017-01/msg00001.html).  I updated my repo accordingly. It now includes the [compiled version](https://github.com/poppyschmo/vim-latexrefman/tree/master/refman).

**Comment by petRUShka** [comment]
@kiryph, thanks for your 50 cents! I think that your links are very useful.

`Zeal` seems to be very interesting solution for offline documentation. But it is less native to VIM then `latexrefman` approach. So I prefer `latexrefman` to be somehow (possible with @poppyschmo plugin if it is ready for `vimtex`) integrated in vimtex.

I'm sure that `zeal` (and `zeavim.vim`) should be mentioned in documentation of vimtex as good offline alternative.

Using `invesigate.vim` for online documentation IMHO much more preferable than reimplementing it in `vimtex`.

**Comment by petRUShka** [comment]
I think that ideal solution is to have `zeal`-like (or `dash`-like) functionality inside vim without neccesity to have some non-vim windows (as zeal itself for example) to be opened.

For example it is impossible to use zeal on non-GUI environments (as ssh-environment is).

**Comment by petRUShka** [comment]
Good news for everyone. There is a console analog for zeal and zeavim.

Meet [dasht](https://github.com/sunaku/dasht) and [vim-dasht](https://github.com/sunaku/vim-dasht).

Possibly that is the solution for in-vim offline documentation (but not for offline).

**Comment by lervag** [comment]
Ok, so I'm ready for this again. After reading the thread, I am thinking the following:

- [ ] Write about offline and online LaTeX related docs in the vimtex doc
- [ ] Add possibility of online search for package manuals like @kiryph suggested

My conclusion is that offline docs seems better provided by Zeal and/or dash(t).

Btw: I'm not getting vim-dasht working. When I try the `Dasht` command, nothing happens and I need to `CTRL-C` to stop an endless process.

**Comment by petRUShka** [comment]
@lervag, are you going to add online documentation with help of [investigate.vim](https://github.com/keith/investigate.vim)?

P.S. I saw your issue in vim-dasht issues. Let's hope author will fix it. In pure console vim it is working well enough.

**Comment by lervag** [comment]
@petRUShka Nah, I don't think so. I will add a simple feature that queries CTAN for package documentation and opens it if found. I might also add a simple command to open the Wikibook. But I don't want to add a direct dependency. I did mention investigate.vim in the update to the docs, though.

By the way, I would be happy to get some constructive feedback on the new help text on LaTeX docs. Feel free to suggest improvements.

**Comment by petRUShka** [comment]
[Link to defaults](https://github.com/keith/investigate.vim/blob/master/autoload/investigate/defaults.vim#L41) for `tex` in investigate.vim

P.S. I'm thinking that dash and company (zeal, dasht) are using `latexrefman` itself, so may be `dasht` is preferable because of more general approach.

**Comment by lervag** [comment]
Thanks! And good point, I think you are right. Although not (for me) until dasht works properly in an unblocking manner. In the meantime, I find zeavim a better alternative.

**Comment by petRUShka** [comment]
I personally don't prefer zeavim over vim-dsaht because of GUI-dependency of zeavim (which is critical in some cases).

**Comment by languitar** [comment]
Just for the reference: linking to http://texdoc.net might also be possible.

**Comment by lervag** [comment]
@languitar Thanks!

I'm curious: How about me adding a simple function that looks up documentation through http://texdoc.net in a browser? It would not be a perfect solution, since we need to escape Vim, but it is simple to implement and it could be a nice addition that would make documentation easier to access for everyone who does not already use e.g. Dasht or Zeavim or similar solutions...

**Comment by petRUShka** [comment]
Possibly pdf with documentation can be sourced offline from `texdoc` utility and in case of absence of documentation on local pc sourced from http://texdoc.net.

**Comment by languitar** [comment]
On archlinux there is no more documentation installed and texdoc would be useless.

**Comment by lervag** [comment]
@languitar But @petRUShka has a good suggestion, I think, in that you can first try offline with `texdoc`, then online with `http://texdoc.net` if no match is found offline.

I think this is a good idea. So, to be precise, I propose the following:

1. Create a command `:VimtexDoc <arg>` that searches first for offline docs with `texdoc`, and if not found, then searches online with `http://texdoc.net` (through ones browser).
2. Create a mapping `K` that runs `:VimtexDoc` on the current keyword.

What do you say?

**Comment by lervag** [comment]
Ok, I've created a minimal implementation that opens texdoc.net. It works on Linux as of now. This is still alpha and has no documentation. `:VimtexDoc word` currently opens texdoc.net/pkg/word. `VimtexDoc` with no argument opens the selected usepackage command's name (assume single package per statement).

Note that `:VimtexDoc not_a_package` will now do a `kpsewhich not_a_package.sty` to determine if the word is a package or not. This is not quite perfect, because the argument can be a package that is not available/installed. However, I was thinking that some kind of check like this could be useful.

I will appreciate suggestions for how this should evolve.

**Comment by andymass** [comment]
@lervag, is the plan to eventually look up which package a command belongs to and then go to that package's documentation?  E.g. on `\binom`, open `texdoc.net/pkg/amsmath`.  Of course, this information is already available through `vimtex#complete#omnifunc`.

By the way, while trying out `VimtexDoc`, I noticed `vimtex#cmd#get_current()` does not work when the cursor is at last character of a command.

**Comment by lervag** [comment]
@andymass That's a good idea! I've implemented this now, please test. I've also fixed the issue with `get_current()`, and I've added some handling of multiple packages, e.g. `\usepackage{p1, p2, p3}`. If ambiguous, the current behaviour is to choose `p1`.

**Comment by andymass** [comment]
Looks pretty good, thanks!  A few things I noticed while testing:
- `vimtex#cmd#get_current()` still returns `{}` at the end of a command e.g. `\bino[m]` where [ ] indicates cursor position.
- If the command is a "built-in," i.e. located in the `complete/default` file, some sort of latex documentation should be brought up.  Two options (among many) are `texdoc.net/pkg/latex2e` and `texdoc.net/pkg/lshort`.
- Vimtex should probably be able to easily handle documentclasses since `texdoc.net/pkg/scrbook` works even though `scrbook` is not a package.
- texdoc (the command) allows one to specify languages other than English in a config file, and you can do this with texdoc.net by appending a language code, e.g., `de`, `texdoc.net/pkg/scrbook-de` for german.  However, it does not default to English, but goes to the main page when documentation in another language cannot be found.

**Comment by lervag** [comment]
Ok, I've handled your three first points. The last is more difficult. I could add an option to change the default url, but it seems silly when most docs are only available in english.

Do you know a simple way to "check" if the documentation exists?

I could use curl, I guess, but then it becomes difficult to make this work on Windows and OSX as well...

**Comment by andymass** [comment]
Points two and three work quite well now- the choice on ambiguous packages is pretty slick!  However, I am still having trouble with point one.  I would be happy to create an issue if you cannot replicate.

I don't know of any way, besides fetching headers with curl/wget, to determine if a particular documentation exists on `texdoc.net`.

One simple approach would be to allow the user to override the `silent execute '!xdg-open..` line with a custom function that takes `(package, command)` as inputs, so if they have installed the local texdoc documentation (which allows you to specify a preference language), they will use that instead.   I agree with you, it's probably not worth putting too much effort into since much/most documentation is English-only anyway (but other users in this thread may be better to comment on this, since I only use English).

**Comment by lervag** [comment]
@andymass Yes, please open a new issue for `get_current()`.

And yes, I like your suggestion for the custom function. I could make a default function for Linux, which uses curl to check if the package is available, something like this should do it:

```sh
curl --head http://texdoc.net/show.php?pkg=name
```

I guess we need an option for the package handler. Perhaps `g:vimtex_doc_handler` is a good name? And the default handlers could be named `vimtex#doc#handler_default_[linux|windows|mac]`.

Right now I am using `kpsewhich` as a filter. However, this assumes that every package or documentclass is installed on the system. That might not be full appropriate. Perhaps this filter should be optional?

I think I will leave the language specific request for now. People who are interested in this should be able to define a custom function based on the one I write. I think they will find everything they need, including relevant curl calls.

**Comment by lervag** [comment]
Ok, I've now added a handler and I've documented the command/mapping and the option. Let me know what you think.

**Comment by andymass** [comment]
I think `g:vimtex_doc_handler` is a fine name, and I see why it was not necessary to split up linux/win/etc into separate functions.  My feeling is that `kpsewhich` is fine for now since adding an option to disable later would be easy. 

The current functionality provided is already amazing and is a great addition! :+1:  I'm wondering about couple things which might be nice: 
 
 1. Could the context (i.e. the value of `l:cmd = vimtex#cmd#get_current()`) be passed to the handler, so I could write a custom handler which, for example, finds the command name in the associated pdf?

 2. Would it be possible to somehow defer the prompt of ['latex2e', 'lshort'] etc, since my doc_handler would like to treat latex built-in commands in a different manner, but currently I am forced to select one before the handler is called.  I'm not sure what the best solution would be.  In cases like `\usepackage{pkga,pkgb}`, having a choice before the handler makes sense.

**Comment by lervag** [comment]
Thanks for the kind words, and for the useful comments! :)

I think it makes sense to pass the context to the handler. In fact, I am making some updates where a more general context is passed and the handler gets even better. I hope to finish this to night, else it might be some time late next week.

**Comment by lervag** [comment]
There. I think this should fix both of your concerns. The example in the `test/issues/655/minivimrc` should provide some pointers in how this new style handler works.

I think this will make it relatively easy to create custom handlers and filters. E.g., one could add a filter that loops through the candidates and uses curl to check if there are docs available in other languages and changes the candidate name if that is the case.

I personally find this feature very useful already, and I think this mostly solves the intended purpose. Do you agree we can close this issue as fixed?

**Comment by lervag** [comment]
By the way: Note that one may easily add a handler that uses `dasht` or `zeal` through a handler to open the docs for commands. Let me know if you need a pointer for this.

@petRUShka @poppyschmo @kiryph @languitar

**Comment by lervag** [comment]
I'll close this for now. Feel free to open related issues/requests or to continue the discussion here.
